### PR TITLE
Pyralid style update

### PR DESCRIPTION
**Pull request by sfarrens**
## TL;DR

- Updated ModOpt package structure to be similar to [Pyralid](https://github.com/sfarrens/pyralid-template)
- Planned new release of ModOpt v1.5.0
- Resolved minor bugs
- Incorporated GPU implementation from #97 

##  Detailed Summary

- Removed Travis-CI and added GitHub actions 
  - Test for compliance with [WPS](https://wemake-python-stylegui.de/en/latest/)
  - Test sphinx docs build
  - Test docstring examples
  - Test distribution is compatible with PyPi
  - Support for [codecov](https://codecov.io/)
- Updated code style and docstrings to be compliant with WPS
   - Changes mostly cosmetic
     - positional argument names
     - internal variable names
     - arrangement of long lines
     - appropriate use of raw strings
   - Some errors suppressed, see `setup.cfg` for justification or todo item for future issue
   - Removed `info.py` and put metadata in `setup.py`
   - Minor API changes
     - changed `Gaussian_filter` to `gaussian_filter` in `filter.py`
     - added `verbose` option to `cwbReweight` in `reweight.py`
     - Removed local imports
- Fixed some broken docstring examples
- Set fixed versions for dependencies to retain stable release
- Fixed minor bugs in `positivity.py` arising from how  recent versions of `numpy` handle `dtype=object`
- Added `notebooks` directory for future tutorials/examples
-  Updated docs and README

## Notes for Reviewers

- This PR is targeting a new `develop` branch. Another PR will be opened to merged this into `master/main` before tagging a release.
- In order to demonstrate the new CI workflow the workflow YAML files have already been pushed to `develop`. This was necessary to trigger GH actions from a fork.
- I would recommend spending most of the reviewing effort checking that the unit tests are OK and nothing important has changed (that is not reported above).

### CODE REVIEW

**Comment by sfarrens** [comment]
Hi @zaccharieramzi and @chaithyagr this PR is finished on my side. There are still some things that could be improved but I will open issues for these. 

Can you review this by next week so that we don't delay the PySAP release by too much? Apologies for the huge PR, but as I explain in the description I think it should be fine if you check that everything is OK with the unit tests. Perhaps you can split the effort between the two of you so that it is not too much hassle for either one?

@chaithyagr please also check that the GPU implementation is OK.

File: modopt/tests/test_opt.py [hunk 1]
```diff
@@ -21,94 +24,147 @@
     import_sklearn = True
 
 
-class dummy(object):
+dummy_func = lambda x_val: x_val
+dummy_func_double = lambda x_val: x_val * 2
+dummy_func_sq = lambda x_val: x_val ** 2
+dummy_func_cube = lambda x_val: x_val ** 3
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:30]: Why `dummy_` as a prefix? I assume this is because you wanted to get rid of unwanted pep8 warnings, but I think that it would be preferable to change the case to BIG rather than adding `dummy_`. It makes the reader think that these functions serve no purpose when they actually do.

File: modopt/tests/test_opt.py [hunk 2]
```diff
@@ -21,94 +24,147 @@
     import_sklearn = True
 
 
-class dummy(object):
+dummy_func = lambda x_val: x_val
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:27]: Maybe specify that this is identity in the variable name

File: modopt/tests/test_opt.py (hunk 1 above)
Comment (zaccharieramzi) [reply modopt/tests/test_opt.py:30]: I see it's also because some of them were actually used as dummy. I think it's even more important to make the difference then.

File: modopt/tests/test_opt.py [hunk 3]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
 
-        npt.assert_equal(self.parent.op(3), 9,
-                         err_msg='Inccoret proximity parent operation.')
-
-        npt.assert_equal(self.parent.cost(3), 6,
-                         err_msg='Inccoret proximity parent cost.')
+        npt.assert_equal(
+            self.parent.cost(3),
+            6,
+            err_msg='Inccoret proximity parent cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:726]:
```suggestion
            err_msg='Incorrect proximity parent cost.',
```

File: modopt/tests/test_opt.py [hunk 4]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+        )
 
     def test_identity(self):
+        """Test IdentityProx."""
+        npt.assert_equal(
+            self.identity.op(3),
+            3,
+            err_msg='Inccoret proximity identity operation.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:734]:
```suggestion
            err_msg='Incorrect proximity identity operation.',
```

File: modopt/tests/test_opt.py [hunk 5]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
 
-        npt.assert_equal(self.identity.op(3), 3,
-                         err_msg='Inccoret proximity identity operation.')
-
-        npt.assert_equal(self.identity.cost(3), 0.0,
-                         err_msg='Inccoret proximity identity cost.')
+        npt.assert_equal(
+            self.identity.cost(3),
+            0,
+            err_msg='Inccoret proximity identity cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:740]:
```suggestion
            err_msg='Incorrect proximity identity cost.',
```

File: modopt/tests/test_opt.py [hunk 6]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
 
-        npt.assert_equal(self.positivity.op(-3), 0,
-                         err_msg='Inccoret proximity positivity operation.')
-
-        npt.assert_equal(self.positivity.cost(-3, verbose=True), 0.0,
-                         err_msg='Inccoret proximity positivity cost.')
+        npt.assert_equal(
+            self.positivity.cost(-3, verbose=True),
+            0,
+            err_msg='Inccoret proximity positivity cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:754]:
```suggestion
            err_msg='Incorrect proximity positivity cost.',
```

File: modopt/tests/test_opt.py [hunk 7]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+            0,
+            err_msg='Inccoret proximity identity cost.',
+        )
 
     def test_positivity(self):
+        """Test Positivity."""
+        npt.assert_equal(
+            self.positivity.op(-3),
+            0,
+            err_msg='Inccoret proximity positivity operation.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:748]:
```suggestion
            err_msg='Incorrect proximity positivity operation.',
```

File: modopt/tests/test_opt.py [hunk 8]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
 
-        npt.assert_array_equal(self.sparsethresh.op(self.data1), self.data2,
-                               err_msg='Inccorect sparse threshold operation.')
-
-        npt.assert_equal(self.sparsethresh.cost(self.data1, verbose=True),
-                         108.0, err_msg='Inccoret sparse threshold cost.')
+        npt.assert_equal(
+            self.sparsethresh.cost(self.data1, verbose=True),
+            108.0,
+            err_msg='Inccoret sparse threshold cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:768]:
```suggestion
            err_msg='Incorrect sparse threshold cost.',
```

File: modopt/tests/test_opt.py [hunk 9]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+            0,
+            err_msg='Inccoret proximity positivity cost.',
+        )
 
     def test_sparse_threshold(self):
+        """Test SparseThreshold."""
+        npt.assert_array_equal(
+            self.sparsethresh.op(self.data1),
+            self.data2,
+            err_msg='Inccorect sparse threshold operation.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:762]:
```suggestion
            err_msg='Incorrect sparse threshold operation.',
```

File: modopt/tests/test_opt.py [hunk 10]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+            108.0,
+            err_msg='Inccoret sparse threshold cost.',
+        )
 
     def test_low_rank_matrix(self):
+        """Test LowRankMatrix."""
+        npt.assert_almost_equal(
+            self.lowrank.op(self.data3),
+            self.data4,
+            err_msg='Inccorect low rank operation: standard',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:776]:
```suggestion
            err_msg='Incorrect low rank operation: standard',
```

File: modopt/tests/test_opt.py [hunk 11]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
-                                err_msg='Inccorect low rank operation: '
-                                        'standard')
-
-        npt.assert_almost_equal(self.lowrank_ngole.op(self.data3), self.data5,
-                                err_msg='Inccorect low rank operation: '
-                                        'ngole')
+        npt.assert_almost_equal(
+            self.lowrank_ngole.op(self.data3),
+            self.data5,
+            err_msg='Inccorect low rank operation: ngole',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:782]:
```suggestion
            err_msg='Incorrect low rank operation: ngole',
```

File: modopt/tests/test_opt.py [hunk 12]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+        )
 
-        npt.assert_almost_equal(self.lowrank.cost(self.data3, verbose=True),
-                                469.39132942464983,
-                                err_msg='Inccoret low rank cost.')
+        npt.assert_almost_equal(
+            self.lowrank.cost(self.data3, verbose=True),
+            469.39132942464983,
+            err_msg='Inccoret low rank cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:788]:
```suggestion
            err_msg='Incorrect low rank cost.',
```

File: modopt/tests/test_opt.py [hunk 13]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+        )
 
     def test_linear_comp_prox(self):
-        npt.assert_array_equal(self.linear_comp.op(self.data1), self.data2,
-                               err_msg='Inccorect sparse threshold operation.')
+        """Test LinearCompositionProx."""
+        npt.assert_array_equal(
+            self.linear_comp.op(self.data1),
+            self.data2,
+            err_msg='Inccorect sparse threshold operation.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:796]:
```suggestion
            err_msg='Incorrect sparse threshold operation.',
```

File: modopt/tests/test_opt.py [hunk 14]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+        )
 
-        npt.assert_equal(self.linear_comp.cost(self.data1, verbose=True),
-                         108.0, err_msg='Inccoret sparse threshold cost.')
+        npt.assert_equal(
+            self.linear_comp.cost(self.data1, verbose=True),
+            108.0,
+            err_msg='Inccoret sparse threshold cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:802]:
```suggestion
            err_msg='Incorrect sparse threshold cost.',
```

File: modopt/tests/test_opt.py [hunk 15]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
     def test_proximity_combo(self):
-
+        """Test ProximityCombo."""
         for data7, data8 in zip(self.data7, self.data8):
-            npt.assert_array_equal(data7, data8,
-                                   err_msg='Inccorect combined operation')
+            npt.assert_array_equal(
+                data7,
+                data8,
+                err_msg='Inccorect combined operation',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:811]:
```suggestion
                err_msg='Incorrect combined operation',
```

File: modopt/tests/test_opt.py [hunk 16]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+            )
 
-        npt.assert_equal(self.combo.cost(self.data6), 0.0,
-                         err_msg='Inccoret combined cost.')
+        npt.assert_equal(
+            self.combo.cost(self.data6),
+            0,
+            err_msg='Inccoret combined cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:817]:
```suggestion
            err_msg='Incorrect combined cost.',
```

File: modopt/tests/test_opt.py [hunk 17]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
-    def test_ridge(self):
+            err_msg='Incorrect sparse threshold operation.',
+        )
 
-        npt.assert_array_equal(self.ridge.op(self.data9), self.data10,
-                               err_msg='Incorect shrinkage operation.')
+        npt.assert_equal(
+            self.owl.cost(self.data1.flatten(), verbose=True),
+            108.0,
+            err_msg='Incorret sparse threshold cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:847]:
```suggestion
            err_msg='Incorrect sparse threshold cost.',
```

File: modopt/tests/test_opt.py [hunk 18]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
-                            err_msg='Incorect sparse Ksupport norm operation',
-                            rtol=1e-6)
-
-        npt.assert_equal(self.five_support_norm.cost(self.data1.flatten(),
-                                                     verbose=True),
-                         684.0, err_msg='Inccoret 3-support norm cost.')
+        npt.assert_equal(
+            self.five_support_norm.cost(self.data1.flatten(), verbose=True),
+            684.0,
+            err_msg='Inccoret 3-support norm cost.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:927]:
```suggestion
            err_msg='Incorrect 5-support norm cost.',
```

File: modopt/tests/test_opt.py [hunk 19]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
+        npt.assert_equal(
+            self.one_support.cost(self.data1.flatten(), verbose=True),
+            259.2,
+            err_msg='Incorect sparse threshold cost.',
+        )
 
-        npt.assert_raises(ValueError, proximity.KSupportNorm, 0.0, 0)
+        npt.assert_raises(ValueError, proximity.KSupportNorm, 0, 0)
 
     def test_three_support_norm(self):
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:915]:
```suggestion
    def test_five_support_norm(self):
```

File: modopt/tests/test_opt.py [hunk 20]
```diff
@@ -498,199 +713,282 @@ def tearDown(self):
...
-
-        npt.assert_almost_equal(self.d_support.cost(self.data9.flatten(),
-                                                    verbose=True),
-                                408.0 * 3.0,
-                                err_msg='Incorect shrinkage cost for' +
-                                        ' d-support norm.')
+        npt.assert_almost_equal(
+            self.d_support.cost(self.data9.flatten(), verbose=True),
+            408.0 * 3.0,
+            err_msg='Incorect shrinkage cost for d-support norm.',
```
Comment (zaccharieramzi) [inline modopt/tests/test_opt.py:944]:
```suggestion
            err_msg='Incorrect shrinkage cost for d-support norm.',
```

**Review by zaccharieramzi — COMMENTED** [review]
Finishing first step in the review. I reviewed the tests, the doc changes, some small files and still have the plot, opt, signal, math modules to review fully, although I am confident most is unchanged thanks to the unit tests (will carry on tmr).

I didn't see any problem standing out and allowed myself a small rampage on "Incorrect" typos (not introduced in this PR but since we are correcting stuff might as well).

Minor points:
- Generally, I think the `'{0}'.format(myvalue)` is not very readable. I think `f'{myvalue}'` is much more seamless.

File: .gitignore [hunk 21]
```diff
@@ -1,21 +1,142 @@
...
+dmypy.json
+
+# Pyre type checker
+.pyre/
+
+# pytype static type analyzer
+.pytype/
 
-# Ignore pypi config
-.pypirc
```
[the remark this answers is not in the document]
Comment (sfarrens) [reply .gitignore:21]: I don't think there is any specific reason to expect that someone will have a `.pypirc` file, I think this is just a precaution in case someone does and accidentally uploads their PyPi login details.

File: .github/workflows/ci-build.yml [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+          python-version: ${{ matrix.python-version }}
+          auto-activate-base: false
+
+      - name: Install Dependencies
+        shell: bash -l {0}
+        run: |
+          python --version
+          python -m pip install --upgrade pip
+          python -m pip install -r develop.txt
+          python -m pip install -r docs/requirements.txt
```
[the remark this answers is not in the document]
Comment (sfarrens) [reply .github/workflows/ci-build.yml]: Good catch, I have removed this.

**Comment by zaccharieramzi** [comment]
Well I think that for the sake of making the process quick we shouldn't bother with this.
It's just that you introduced some `'{0}'.format(myvalue)` statements in this PR, that's why I mentioned it.

**Comment by sfarrens** [comment]
Indeed, I would like to keep everything consistent (at least as much as possible). So perhaps I can open an issue to replace `format` statements with f-strings for the following release (mid 2021).

File: modopt/opt/algorithms.py [hunk 23]
```diff
@@ -78,50 +80,123 @@ class SetUp(Observable):
...
+                name,
+                dic['metric'],
+                dic['mapping'],
+                dic['cst_kwargs'],
+                dic['early_stopping'],
+            )
+            self.add_observer('cv_metrics', observer)
+
+        # Check for GPU
+        if use_gpu:
```
Comment (zaccharieramzi) [inline modopt/opt/algorithms.py:127]: It seems weird to me that some GPU functionalities are introduced in this PR which should only deal with style right?

Comment (sfarrens) [reply modopt/opt/algorithms.py:127]: Hi @zaccharieramzi, @chaithyagr pushed to include the changes he has implemented in Incorporated #97. Since these don't (or at least as far I could see) impact the standard CPU implementation or API I figured it should be OK for now.

I will need to open an issue to standardise this throughout the code, but I think this could also be done for the mid 2021 release.

If there is anything specific that you are concerned about please point it out. I would rather delay the release by a couple of days to sort it out rather than rushing through something that could cause problems later on.

File: modopt/math/matrix.py [hunk 24]
```diff
@@ -315,30 +347,34 @@ def get_spec_rad(self, tolerance=1e-6, max_iter=20, extra_factor=1.0):
...
         # Iterate until the L2 norm of x converges.
-        for i in range(max_iter):
+        for i_elem in range(max_iter):
 
             x_old_norm = np.linalg.norm(x_old)
 
             x_new = self._operator(x_old) / x_old_norm
 
-            x_new_norm = np.linalg.norm(x_new)
+            xp = get_array_module(x_new)
```
Comment (chaithyagr) [inline modopt/math/matrix.py:360]: I am unable to do a suggestion here, this must be:
```python
            xp = get_array_module(x_new)

            x_old_norm = xp.linalg.norm(x_old)

            x_new = self._operator(x_old) / x_old_norm

```

**Review by chaithyagr — CHANGES_REQUESTED** [review]
Minute quick change,
Else look fine, you can merge once this is done and when green. 
I think we need another issue to have atleast a GPU pipeline of test, which we can run locally. @sfarrens can you add me on that? I will try to run the same tests on GPU. 
For now, this PR is already too bulky to handle this.

File: modopt/math/matrix.py (hunk 24 above)
Comment (sfarrens) [reply modopt/math/matrix.py:360]: I guess you mean

```python
xp = get_array_module(x_old)

x_old_norm = xp.linalg.norm(x_old)

x_new = self._operator(x_old) / x_old_norm

x_new_norm = xp.linalg.norm(x_new)
```

as `x_new` is not defined prior to the call to `get_array_module`.

Comment (chaithyagr) [reply modopt/math/matrix.py]: Oops, yes you are right!

**Comment by zaccharieramzi** [comment]
@sfarrens I think you didn't see some of the comments I made during my first PR because they are collapsed by the GitHub viewer

File: modopt/tests/test_opt.py (hunk 2 above)
Comment (sfarrens) [reply modopt/tests/test_opt.py]: renamed to `func_identity`

File: modopt/tests/test_opt.py (hunk 1 above)
Comment (sfarrens) [reply modopt/tests/test_opt.py]: Yeah, fair point. I have removed the `dummy_` prefix. `func` on it's own would raise errors but with the suffixes everything is fine. I also added a comment line to explain what these lambda functions are for.

**Comment by sfarrens** [comment]
Thanks @zaccharieramzi, I think I got everything now. 😅
